File: scripts/generate-brand-assets.php
<?php

/**
 * Generate favicons, site logo, and OG default image from source PNGs.
 *
 * Sources (project root):
 *   logo2026.png     → favicons + public/logo.png
 *   logo_fill.png   → OG default image logo tile
 */

$root = dirname(__DIR__);
$public = $root . '/public';
$imagesDir = $public . '/images';

$logoSource = $root . '/logo2026.png';
$ogLogo = $public . '/logo_fill.png';

foreach ([$logoSource, $ogLogo] as $file) {
    if (! is_file($file)) {
        fwrite(STDERR, "Missing source file: {$file}\n");
        exit(1);
    }
}

if (! is_dir($imagesDir)) {
    mkdir($imagesDir, 0755, true);
}

function loadPng(string $path): GdImage
{
    $image = imagecreatefrompng($path);
    if ($image === false) {
        throw new RuntimeException("Could not load PNG: {$path}");
    }

    imagealphablending($image, true);
    imagesavealpha($image, true);

    return $image;
}

function resizeSquare(GdImage $source, int $size): GdImage
{
    $canvas = imagecreatetruecolor($size, $size);
    imagealphablending($canvas, false);
    imagesavealpha($canvas, true);

    $transparent = imagecolorallocatealpha($canvas, 0, 0, 0, 127);
    imagefill($canvas, 0, 0, $transparent);

    $width = imagesx($source);
    $height = imagesy($source);
    imagecopyresampled($canvas, $source, 0, 0, 0, 0, $size, $size, $width, $height);

    return $canvas;
}

function savePng(GdImage $image, string $path): void
{
    if (! imagepng($image, $path)) {
        throw new RuntimeException("Could not write PNG: {$path}");
    }
}

$logo = loadPng($logoSource);
copy($logoSource, $public . '/logo.png');

foreach ([16 => 'favicon-16x16.png', 32 => 'favicon-32x32.png', 180 => 'apple-touch-icon.png'] as $size => $filename) {
    $resized = resizeSquare($logo, $size);
    savePng($resized, $public . '/' . $filename);
    imagedestroy($resized);
    echo "Wrote {$filename} ({$size}x{$size})\n";
}

$icoScript = sprintf(
    "from PIL import Image\nimg = Image.open(r'%s')\nimg16 = img.resize((16, 16), Image.Resampling.LANCZOS)\nimg.save(r'%s', format='ICO', sizes=[(16, 16), (32, 32)])\n",
    $public . '/favicon-32x32.png',
    $public . '/favicon.ico'
);
$proc = proc_open('python -', [['pipe', 'r'], ['pipe', 'w'], ['pipe', 'w']], $pipes);
if (is_resource($proc)) {
    fwrite($pipes[0], $icoScript);
    fclose($pipes[0]);
    proc_close($proc);
    echo "Wrote favicon.ico\n";
}

$ogPy = $root . '/scripts/generate-og-image.py';
$tmpOg = $public . '/og-default-tmp.png';
$pyCmd = sprintf(
    'python %s %s %s',
    escapeshellarg($ogPy),
    escapeshellarg($ogLogo),
    escapeshellarg($tmpOg)
);

$pyOk = false;
exec($pyCmd . ' 2>&1', $pyOut, $pyCode);
if ($pyCode === 0 && is_file($tmpOg)) {
    foreach (['og-default.png', 'images/og-default.png'] as $relative) {
        copy($tmpOg, $public . '/' . $relative);
        echo "Wrote {$relative} (clean headline layout)\n";
    }
    unlink($tmpOg);
    $pyOk = true;
}

if (! $pyOk) {
    fwrite(STDERR, "Python OG generation failed. Install Pillow and run:\n");
    fwrite(STDERR, "  python scripts/generate-og-image.py public/logo_fill.png public/og-default.png\n");
    if (! empty($pyOut)) {
        fwrite(STDERR, implode("\n", $pyOut) . "\n");
    }
    imagedestroy($logo);
    exit(1);
}

imagedestroy($logo);

echo "Done.\n";

File: scripts/generate-brand-assets.py
from PIL import Image, ImageDraw, ImageFont
import os

SRC = os.path.join(os.path.dirname(__file__), "..", "logo2026.png")
PUBLIC = os.path.join(os.path.dirname(__file__), "..", "public")
FAV = os.path.join(PUBLIC, "favicons")
os.makedirs(FAV, exist_ok=True)

logo = Image.open(SRC).convert("RGBA")


def fit_square(size, padding=0.12):
    canvas = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    inner = int(size * (1 - padding * 2))
    copy = logo.copy()
    copy.thumbnail((inner, inner), Image.Resampling.LANCZOS)
    x = (size - copy.width) // 2
    y = (size - copy.height) // 2
    canvas.paste(copy, (x, y), copy)
    return canvas


main = logo.copy()
main.thumbnail((512, 512), Image.Resampling.LANCZOS)
main.save(os.path.join(PUBLIC, "logo.png"), optimize=True)

for size, name in [
    (16, "favicon-16x16.png"),
    (32, "favicon-32x32.png"),
    (180, "apple-touch-icon.png"),
]:
    fit_square(size).save(os.path.join(PUBLIC, name), optimize=True)

for size, name in [(16, "icon-16.png"), (32, "icon-32.png"), (192, "icon-192.png"), (512, "icon-512.png")]:
    fit_square(size).save(os.path.join(FAV, name), optimize=True)

fit_square(256).save(
    os.path.join(PUBLIC, "favicon.ico"),
    format="ICO",
    sizes=[(16, 16), (32, 32), (48, 48), (64, 64)],
)

w, h = 1200, 630
og = Image.new("RGB", (w, h))
draw = ImageDraw.Draw(og)
for y in range(h):
    t = y / h
    r = int(41 + (45 - 41) * t)
    g = int(121 + (55 - 121) * t)
    b = int(255 + (204 - 255) * t)
    draw.line([(0, y), (w, y)], fill=(r, g, b))

og_logo = logo.copy()
og_logo.thumbnail((320, 320), Image.Resampling.LANCZOS)
lx = (w - og_logo.width) // 2
ly = (h - og_logo.height) // 2 - 40
og_rgba = og.convert("RGBA")
og_rgba.paste(og_logo, (lx, ly), og_logo)

try:
    font = ImageFont.truetype("arialbd.ttf", 52)
except OSError:
    font = ImageFont.load_default()

draw2 = ImageDraw.Draw(og_rgba)
text = "Koding Indonesia"
bbox = draw2.textbbox((0, 0), text, font=font)
tx = (w - (bbox[2] - bbox[0])) // 2
ty = ly + og_logo.height + 24
draw2.text((tx + 3, ty + 3), text, fill=(0, 0, 0, 180), font=font)
draw2.text((tx, ty), text, fill=(255, 255, 255, 255), font=font)
og_rgba.convert("RGB").save(os.path.join(PUBLIC, "og-default.png"), optimize=True)

print("Generated brand assets OK")
